plugins/api/api_main.ml

```diff
@@ -43,37 +103,95 @@ let main proj =
+      Format.print_newline () in
+  let apis = api_of_paths paths |>
+             List.sort ~cmp:(fun x y -> String.compare x.lang y.lang) in
+  List.iter ~f:print_lang (List.group ~break:(fun x y -> x.lang <> y.lang) apis);
+  Format.printf
+    "Total number of available API: %d\n" (List.length apis)
+
+let rec add_files apis paths = match paths with
+  | [] ->
+    eprintf "There are not any writable paths for api plugin\n";
+    exit 1
+  | path :: paths' ->
+    match apis with
+    | [] -> ()
+    | api :: apis' ->
+      match Api_path.cp path api with
+      | Ok () -> add_files apis' paths
+      | Error `No_source ->
+        eprintf "Api %s does not exist\n" api.file;
+        add_files apis' paths
+      | Error `Duplicated ->
+        eprintf "Api %s is already exists in api paths (%s), not added\n"
+          api.file path;
+        add_files apis' paths
+      | Error (`Fail reason) ->
+        eprintf "Failed to copy %s to %s because of error: %s\n"
+          api.file path reason;
+        add_files apis paths'
+
+let rem_files apis paths =
+  let rm api =
+    let found = List.fold ~init:[] ~f:(fun acc path ->
+        match Api_path.rm path api with
+        | Ok () -> path :: acc
+        | Error (`Fail reason) ->
+          eprintf "Api %s can't be removed from %s because of error %s\n"
+            api.file path reason;
+          path :: acc
+        | Error `No_source -> acc) paths in
+    if List.is_empty found then
+      eprintf "Api %s was not found in paths\n" api.file in
+  List.iter ~f:rm apis
 
-let add_files api =
-  let bundle = main_bundle () in
-  List.map api ~f:(fun api ->
-      Some ("api/" ^ api.lang ^ "/" ^ Filename.basename api.file),
-      Uri.of_string api.file) |>
-  Bundle.insert_files bundle
-
-let rem_files api =
-  main_bundle () |>
-  Bundle.update ~f:(fun file -> match String.split file ~on:'/' with
-      | ["api"; lang; file] ->
-        if List.exists api ~f:(fun x ->
-            x.lang = lang && x.file = Filename.basename file)
-        then `Drop
-        else `Copy
-      | "api" :: _ ->
-        warning "removing from the bundle a malformed entry: %s" file;
-        `Drop
-      | _ -> `Copy)
+let sanity_check paths =
+  let is_same api api' =
+    api.file <> api'.file && api.lang = api'.lang &&
+    Filename.basename api.file = Filename.basename api'.file in
+  let check_for_pair api apis =
+    match List.find ~f:(is_same api) apis with
+    | None -> ()
+    | Some api' ->
+      eprintf "sanity check failed for api plugin, apis with same \
+               names exists for %s language in %s and %s\n"
+        api.lang api.file api'.file;
+      exit 1 in
+  let all = api_of_paths paths in
+  List.iter ~f:(fun api -> check_for_pair api all) all
 
+let check_paths paths =
+  List.filter_map paths
+    ~f:(fun p ->
+        if Api_path.exists p then Some p
+        else
+          let () = eprintf
+              "Api path %s doesn't exists, excluded from paths\n" p in
```

Comment:
First of all it is a warning, i.e., an error condition that doesn't impend further actions. Second, I'm not sure that we should actually make noise about it. Consider, how gcc (or any other compiler)( acts when you specify  `-I<dir>` and `<dir>` doesn't exist? So probably, we should just ignore it.